File: python3/app/modules/summonapi.py
#!/usr/bin/python3
import base64
import hashlib
import hmac
from urllib.parse import urlencode, quote_plus, unquote_plus
from datetime import datetime
import requests


def search(access_id, api_key, params):
    """
    Searches the Summon API with the provided parameters.

    :param access_id: Your Summon API Access ID
    :param api_key: Your Summon API Key
    :param params: Search parameters, sent as dictionary or list of tuples
    :returns: JSON response from Summon API containing search results, etc.
    """
    host = 'api.summon.serialssolutions.com'
    path = '/2.0.0/search'
    query = urlencode(params, doseq=True, quote_via=quote_plus)

    headers = build_headers(access_id, api_key, host, path, query)
    print('Request headers: {}'.format(headers))

    url = 'http://{}{}?{}'.format(host,
                                  path,
                                  query)
    response = requests.get(url, headers=headers)
    return response.text


def build_headers(access_id, api_key, host, path, query):
    """
    Generates the request headers for the Summon API query.

    :param access_id: Your Summon API Access ID
    :param api_key: Your Summon API Key
    :param host: The Summon API host domain, defined in search function
    :param path: The Summon API search path, defined in search function
    :param query: URL-encoded query, created from params by search function
    :returns: Dictionary containing request headers for Summon API query
    """
    # application/xml also supported by Summon API but not used in this demo
    accept = 'application/json'
    date = datetime.utcnow().strftime('%a, %d %b %Y %H:%M:%S GMT')

    # sort and decode query
    query_string = unquote_plus('&'.join(sorted(query.split('&'))))

    id_string = "\n".join([accept, date, host, path, query_string]) + "\n"

    auth_string = build_auth_string(access_id, api_key, id_string)

    return {'Accept': accept,
            'x-summon-date': date,
            'Host': host,
            'Authorization': auth_string}


def build_auth_string(access_id, api_key, id_string):
    """
    Generates authentication string needed for Authorization header.

    :param access_id: Your Summon API Access ID
    :param api_key: Your Summon API Key
    :param id_string: string from build_headers combining query, etc.
    :returns: string "Summon your_access_id;base64_encoded_hash"
    """
    key = bytes(api_key, 'UTF-8')
    message = bytes(id_string, 'UTF-8')
    hashed_code = hmac.new(key,
                           message,
                           hashlib.sha1).digest()
    digest = base64.encodebytes(hashed_code).decode('UTF-8')

    auth_string = 'Summon {};{}'.format(access_id, digest)
    return auth_string.replace('\n', '')

File: node.js/summonapi.js
console.log('test');
var crypto = require('crypto');
var http = require('http');
var querystring = require('querystring');
var url = require('url');
var util = require('util');

//add summon api credentials as string. TODO: replace with environment variables
var accessID;
var key;

var host = 'api.summon.serialssolutions.com';


http.createServer(function (request, response) {
  util.log(request.method + ' http://' + request.headers.host + request.url);
  var urlParts = url.parse(request.url);
  var accept = 'application/json';
  var date = (new Date()).toGMTString();
  var path = '/2.0.0/search';
  //TODO: Retrieve form parameters for query
  var query = 's.q=test';

  //TODO: Separate sections below into functions

  var unQuery;
  if (urlParts.query) {
    var queryParts = urlParts.query.split('&');
    var params = [];
    for (var i = 0; i < queryParts.length; i++) {
      var param = queryParts[i];
      if (param && (param.indexOf('=') != -1)) {
        param = param.replace('+', '%20');
        params.push(param);
      }
    }
    params.sort();
    query = params.join('&');
    unQuery = querystring.unescape(query);
  } else {
    query = unQuery = '';
  }

  var idString = accept + '\n' + date + '\n' + host + '\n' + path + '\n' + unQuery + '\n';
  var hmac = crypto.createHmac('sha1', key);
  var hash = hmac.update(idString);
  var digest = hash.digest('base64');
  var headers = {
    accept: accept,
    'x-summon-date': date,
    'Host': host,
    'Authorization': 'Summon ' + accessID + ';' + digest
  };


  var options = {
    hostname: host,
    port: 80,
    path: path,
    method: 'GET',
    headers: headers
  }
  var summonRequest = http.request(options);
  summonRequest.end();
  summonRequest.on('response', function(summonResponse) {
    if (summonResponse.statusCode !== 200) {
      response.writeHead(summonResponse.statusCode, 
          {'Content-Type': 'text/html'});
    } else {
      response.writeHead(200, {'Content-Type': 'application/json'});
    }
    summonResponse.setEncoding('utf8');
    summonResponse.on('data', function(chunk) {
      response.write(chunk);
    });
    summonResponse.on('end', function() {
      response.end();
    });
  });
}).listen(8124);

console.log('Server running at http://127.0.0.1:8124/');